stablehlo/transforms/StablehloRefineShapes.cpp

```diff
@@ -0,0 +1,937 @@
+    auto value = std::get<0>(it);
+    auto refinedType = std::get<1>(it);
+    if (value.getType() == refinedType) continue;
+
+    // Check whether the users of this value are ready for the type of the
+    // value to be refined.
+    for (Operation* user : value.getUsers()) {
+      // CHLO and StableHLO ops are designed to support type refinements of
+      // their operands and results. Any operand type in these ops can change
+      // within what's supported by `inferMostSpecificType` without breaking
+      // verification of the op.
+      if (isa<chlo::ChloDialect, StablehloDialect>(user->getDialect()))
+        continue;
+
+      // Simply changing operand type of `func.return` won't work because
+      // that won't update the FunctionType of the enclosing `func.func`.
+      // Nonetheless, we still want to support these ops because they are widely
+      // used in StableHLO programs (although the plan of record is to replace
+      // `func.return` ops in StableHLO programs with `stablehlo.return`:
+      // https://github.com/openxla/stablehlo/issues/425).
+      if (isa<func::ReturnOp>(user)) continue;
+
+      // Unlike in TensorFlow's type inference pass, here we work only with
+      // allowlisted ops to focus our support on well-defined semantics of
+      // StableHLO programs.
+      return rewriter.notifyMatchFailure(op, [&](Diagnostic& diag) {
+        diag << "unsupported refinement: tried to refine " << value.getType()
+             << " to " << refinedType << " for user " << user;
+      });
+    }
+
+    // Happy path: simply call setType here because most of our users are
+    // fine with that.
+    auto unrefinedType = value.getType();
+    value.setType(refinedType);
+
+    // Special case: for `func.return`, guard the refinement with a
+    // `tensor.cast` and leave propagation of the refined return type to a
+    // dedicated pattern.
+    auto isFuncReturn = [](OpOperand& use) -> bool {
+      return isa<func::ReturnOp>(use.getOwner());
+    };
+    if (llvm::none_of(value.getUses(), isFuncReturn)) continue;
+    rewriter.setInsertionPointAfter(op);
+    auto castToUnrefinedType =
+        rewriter.create<tensor::CastOp>(op->getLoc(), unrefinedType, value);
+    value.replaceUsesWithIf(castToUnrefinedType, isFuncReturn);
+  }
+
+  return success();
+}
+
+// Refines the return types of the given operation using the given types.
+// This function also signals PatternRewriter that it needs to visit all the
+// users of this op if any updates to its results have happened during execution
+// of the function.
+LogicalResult refineReturnTypes(PatternRewriter& rewriter, Operation* op,
+                                ArrayRef<Type> types) {
+  if (failed(refineValues(rewriter, op, op->getResults(), types)))
+    return failure();
+
+  // This `replaceOpWithIf` call doesn't actually change the IR, but
+  // it does ask the rewriter to visit all the users of this op. I haven't
+  // found a direct way of doing that.
```

Comment:
> "I haven't found a direct way of doing that."

How about we say "It could be optimized if there's more direct way of doing this."

Reply:
Reworded the comment.